rfd/0014-session-2FA.md

```diff
@@ -0,0 +1,486 @@
+
+#### kubernetes
+
+`kubectl` is configured to call `tsh kube credentials` as an exec plugin, since
+5.0.0. This plugin returns a private key and cert to `kubectl`, which uses them
+in mTLS handshake.
+`tsh kube credentials` will handle the U2F handshake, and cache the resulting
+certificate in `~/.tsh/` for its validity period.
+
+```sh
+$ kubectl get pods
+please tap your security key... <tap>
+... list of pods ...
+
+$ kubectl get pods # no MFA needed right after the previous command
+... list of pods ...
+
+$ sleep 1m && kubectl get pods # MFA needed since the short-lived cert expired
+please tap your security key... <tap>
+... list of pods ...
+
+```
+
+#### web apps
+
+Web apps already have a session concept, with dedicated a login endpoint
+(`/x-teleport-auth`). The application endpoint serves a bit of JS code to
+redirect to the login endpoint.
+
+This JS code will be modified to trigger browser's native U2F API, if the proxy
+responds with a U2F challenge:
+
+- user opens `app.example.com` (with an existing Teleport cookie)
+- proxy serves a minimal JS page
+- JS requests `app.example.com/x-teleport-auth`
+- proxy responds with [407 Proxy Authentication
+  Required](https://tools.ietf.org/html/rfc7235#section-3.2) and a U2F
+  challenge in `Proxy-Authenticate` header
+- JS triggers the browser U2F API
+- browser shows a security key popup
+- user taps the key
+- JS requests `app.example.com/x-teleport-auth` with the signed U2F challenge
+  in `Proxy-Authenticate` header
+- proxy sends back an application-specific cookie and redirects to the
+  application
+
+#### DB
+
+The initial integration for databases will be limited:
+
+```sh
+$ tsh db login prod
+please tap your security key... <tap>
+
+$ eval $(tsh db env)
+$ psql -U awly prod
+```
+
+We'll also provide an example wrapper script:
+
+```sh
+$ cat teleport/examples/db/psql.sh
+#!/bin/sh
+# simplified version, without checking arguments
+
+# Usage: psql.sh user dbname
+tsh db login $2
+eval $(tsh db env)
+psql -U $1 $2
+```
+
+Users will need to adapt this for their DB clients. Teleport will always
+generate short-lived key/cert in a predictable location under `~/.tsh/`.
+
+### API
+
+The protocol to obtain a new cert after a U2F check is:
+```
+client                               server
+   |<-- mTLS using regular tsh cert -->|
+   |--------- initiate U2F auth ------>|
+   |<------------ challenge -----------|
+   |---- u2f signature + metadata ---->|
+   |<-------------- cert --------------|
+```
+
+This can be implemented as 2 request/response round-trips of the existing
+`GenerateUserCerts` RPC, with some downsides:
+- the server has to store state (challenge) in the backend
+- extra latency (backend RTT and RPC overhead)
+- complicating the existing RPC semantics
+
+Instead, we'll use a single _streaming_ gRPC endpoint, using `oneof`
+request/response messages.
+
+```protobuf
+rpc GenerateUserCertMFA(stream UserCertsMFARequest) returns (stream UserCertsMFAResponse);
```

Comment:
Just adding a reminder here. During security testing, we should evaluate the overall state machine implementation and  make sure that the _UserCert_ is issued after challenge verification.